### ISSUE TITLE
Added: GitHub Actions for automated tests

### ISSUE THREAD

**Issue by ghost** [backdrop/backdrop-issues#4753]
There has been discussion about using [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions) for automated testing of PRs, instead of the current Zen.ci service provided by @Gormartsen.

Opening this issue for discussion and possible testing of the service, so we can make a decision if it's going to be suitable for us or not.

UPDATE: there seems to be consensus now, that switching to GHA is the way to go.

**Comment by klonos** [comment]
GitHub Actions have been mentioned a lot in #3213 but perhaps better to have a dedicated issue for that (?) 🤷

**Comment by BWPanda** [comment]
@klonos This issue is specifically for tests. Coding standards I think can be addressed separately. Ultimately they'll hopefully all become part of the same system that each PR is checked against, but since coding standards don't exist yet, then trying to combine them herefirst might be a challenge...

**Comment by klonos** [comment]
> @klonos This issue is specifically for tests.

Yup, I get that and I agree. See how I said:

> ...perhaps better to have a dedicated issue for that

**Comment by BWPanda** [comment]
@jenlampton mentioned during the [dev meeting](https://youtu.be/_7jEjLMTdu4?t=377) that this could be built and tested in the security issue queue where it's not public and won't affect current tests before it's ready and we've decided to switch over.

My concern with that suggestion is that it'll limit who can work on and/or test this, since the security queue is access-only to members of the security team (if I understand that correctly)...

**Comment by quicksketch** [comment]
I adapted a GitHub Action script from a client project to run against Backdrop core. You can see the in-progress results at the PR at https://github.com/backdrop/backdrop/pull/3620

I set this up to get an idea of the general run-time we can expect from GitHub Actions. The GitHub free servers are not incredibly powerful ([see specs here](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners)), only a 2-core CPU, which probably means we won't be able to crank up the concurrency of tests very high, because it's dependent on having multiple cores to really be very effective. Zen.CI currently runs at a concurrency of 10 tests at a time, with the free runners we'll probably only be able to run 1 or 2; adding more concurrency will probably just result in an overall slower result.

@hosef has mentioned previously that another reason Zen.CI runs so quick is because it can use MyISAM (instead of InnoDB) for its database, which is no longer supported on new MySQL installations. So it's likely that no matter what provider we choose (if it's not GitHub Actions) will be slower just because we need to update MySQL to a newer version.

Test runs still haven't finished as of yet, but it looks like it will very likely be more than 30 minutes for a test run. GitHub's timeout is 6 hours though, so at least we won't be hitting up against that. I may experiment with running against a custom self-hosted runner to see if that has the possibility of substantially increasing the speed of things.

**Comment by BWPanda** [comment]
From my quick & dirty research, it seems Aria is the modern equivalent of MyISAM in MariaDB: https://mariadb.com/resources/blog/storage-engine-choice-aria/ Could we use that instead...?

**Comment by quicksketch** [comment]
Thanks @BWPanda, that's a great suggestion. I'm still sorting out the benchmarking of how fast things run. Looks like the ARIA engine exists by default on GitHub runners, which simplifies things, and just a quick flipping of "Engine=MyISAM" to "Engine=ARIA" seems to work in our testing scripts. It looks like a promising alternative at first glance.

**Comment by quicksketch** [comment]
> Which version of MariaDB do GitHub runners use?

Looks like it's actually MySQL 8.0.23 (on Ubuntu 20.04). Although we could really probably run anything we wanted if we brought custom images into the mix. Right now I'm just using the MySQL/Apache/PHP directly on the runner (which is already a container itself).

Here are some timings based on my tests so far:

| Concurrency | InnoDB Engine | ARIA Engine | InnoDB on SHM | ARIA on SHM |
| --- | --- | --- | --- | --- |
| 1 | ?? | ?? | ?? | ?? |
| 4 | ?? | 37.5 min | 31 min | 32 min |
| 10 | ?? | ?? | ?? | ?? |
 
I'll fill in more as results come in.

Supposedly if we use a memory partition (SHM) for MySQL's data storage, that makes InnoDB run comparable to MyISAM. Since these databases are ephemeral, that seems like an obvious enhancement we could use. See http://serebrov.github.io/html/2012-12-17-unit-speed-mysql-to-mem.html

**Comment by indigoxela** [comment]
It might be possible to start more than one runner - they run in parallel. Given above speed, if we start four of them (each runs the tests for 9 to 10 core modules), the tests are done in about 8 minutes - possibly faster than with Zen.CI.
And hopefully without annoying exceptions and the usual turn-it-off-and-on again...

I'm not sure if that's feasible or worth the effort. Maybe we wait until the pain with Zen exceeds the limits. :grin: 

Some background info for GH runners:

- 2 CPU cores
- Intel(R) Xeon(R) Platinum 8171M CPU @ 2.60GHz
- 7 GB of ram

**Comment by indigoxela** [comment]
I also did some experiments, see https://github.com/indigoxela/backdrop/actions

One thing I discovered: 3 fractions seem OK, more seem to cause random failures with installer.test

The upper limit for us _by Github_ would be 20 concurrent runners, BTW.

Here's a screenshot, that run took about 12 Minutes (php 5.3 was slow, 7.4 was a lot faster).

![github-action-summary](https://user-images.githubusercontent.com/842066/124793367-3421b700-df4e-11eb-941d-e30c1472ad69.png)

That failure is BootstrapPageCacheTestCase - it might have something to do with timezone settings.

**Comment by quicksketch** [comment]
I've been researching if it's possible to consolidate all the jobs into a single workflow, and so far I've only found a project call "Check Group" that removes selected workflow jobs so that they don't flood the workflow queue.

So instead of a giant list of workflows like this:

![image](https://user-images.githubusercontent.com/100206/125151636-6c66f800-e0fc-11eb-865b-f59939a63881.png)

We could instead have something like this:

![image](https://user-images.githubusercontent.com/100206/125151714-e7301300-e0fc-11eb-9ffb-a5b8d8a6bebe.png)

Where all the Functional Tests are combined together into a single line item. It then consolidates them all into a single set result that has a sub-page:

![image](https://user-images.githubusercontent.com/100206/125151626-59542800-e0fc-11eb-92d6-b9ce10050cb6.png)

However, I'd like to be able to change the label on the overall workflow, which isn't yet possible (instead it always has the label "Check Group". I filed a feature request against the App at https://github.com/tianhaoz95/check-group/issues/457

**Comment by indigoxela** [comment]
Just in case people aren't aware: the `--split` command line option my approach utilizes to split the tests into chunks already exists for years. :smirk: 

Summary of todos and show-stoppers I currently see before we can switch to Github actions:

1. The biggest showstopper seems the *always* [failing BootstrapPageCacheTestCase](https://github.com/backdrop/backdrop-issues/issues/5126)
2. We should also revisit the other [randomly failing candidates](https://github.com/backdrop/backdrop-issues/issues/1478)
3. Somebody should figure out, how this "Check Group" thing works - I wasn't able to :shrug: 
4. We need an additional command line switch in run-tests.sh to turn off the conversion to MyISAM - see also [this issue](https://github.com/backdrop/backdrop-issues/issues/4392)
5. If my experimental branch is considered for a PR, I'd need to reduce it from "always fires" to only on "PR against 1.x" (easy)
6. ^^ Also, some polish for better readability wouldn't hurt, maybe

**Comment by quicksketch** [comment]
> Somebody should figure out, how this "Check Group" thing works - I wasn't able to shrug

Looks like you may have gotten it... maybe it only groups them if there are no failures? I'll take a look at the failing boostrap page cache test over in https://github.com/backdrop/backdrop-issues/issues/5126

**Comment by indigoxela** [comment]
So I've finally created a [PR](https://github.com/backdrop/backdrop/pull/3688). Note: it's still blocked until the [HTTP handling](https://github.com/backdrop/backdrop-issues/issues/5126) got fixed, but now we have something more concrete to discuss about.

The grouping thing is still unsolved. I tried all sorts of variants, but the grouping plugin simply doesn't seem to work with matrices. 

We have two bottlenecks to consider:

1. The limit of available runners per project (20 for us)
2. The checkout limit (unknown)

Suggestion for 1: we should not start more than the half of available runners for one PR, so two simultaneous PRs don't block each other.

The second one is trickier. When doing my experiments with a branch, I realized that if I started too many runners at the same time, they ran into a timeout somewhere - I assume the checkout didn't work. 6 concurrent runners seem to be a good compromise. Currently Zen.CI jobs still run - so even more checkouts.

And the [randomly failing tests](https://github.com/backdrop/backdrop-issues/issues/1478) are still ... randomly failing.

Several tasks until we can switch to GHA. :wink:

**Comment by quicksketch** [comment]
I'm completely in favor of separating out the grouping functionality until after we have GHA up and running.

I'm not understanding what yo mean by checkout limit. Is that referring to the the size of the repo? Or the number of times a repository can be cloned? I don't think that GitHub would create a limit on the number of times GHA could check out the repository per minute, but then again I've never seen a situation where I tried rapidly checking out the same repository repeatedly.

**Comment by indigoxela** [comment]
> ... what yo mean by checkout limit. Is that referring to the the size of the repo? Or the number of times a repository can be cloned?

The latter - I can't prove it (no real evidence), but ran into this several times when I tried to increase the number of runners to lower the overall time.

**Comment by klonos** [comment]
BTW @quicksketch, the developer in https://github.com/tianhaoz95/check-group/issues/457 has replied, and it seems that custom workflow labels are available now 🎉

**Comment by indigoxela** [comment]
@quicksketch many thanks for merging the blocking fix, I rebased the PR and these tests are passing now.

One of the random failures strikes again: 
```
Fail      Other      locale.test        895 LocaleImportFunctionalTest->testAut
Fail      Other      locale.test        904 LocaleImportFunctionalTest->testAut

    Trying to access array offset on value of type bool
    Trying to access array offset on value of type bool
```
But in one of the runners, there's also that timeout problem again.
I have to correct myself - it has nothing to do with the checkout itself. Backdrop installs and then... nothing happens until the timeout is reached. This needs some more investigation.

**Comment by indigoxela** [comment]
Hehe, out of curiosity I compared the passed fraction with the stalled one.

The test that didn't complete (no entry) in the stalled is one of our "random failure buddies": `Batch API: Batch progress page (BatchPageTestCase)`

So I guess, our next steps should be to make LocaleImportFunctionalTest and BatchPageTestCase more robust.

**Comment by klonos** [comment]
I'm really glad to see progress in this! ...I'm looking forward to us minimizing random failures, and having the ability to manage php versions that the tests run, and eventually add linters in the mix too. Exciting times!

**Comment by indigoxela** [comment]
Tadaa, we have a ... random pass ... :laughing: https://github.com/backdrop/backdrop/actions/runs/1176818518

**Comment by klonos** [comment]
I just wanted to say that 10min total runtime for the tests is absolutely acceptable. I don't mind the 3-5min increase compared to the time I am wasting with Zen CI each time it decides to start having an attitude 😓

**Comment by quicksketch** [comment]
@indigoxela Excellent :smile: 

Could you restore the Zen.CI scripts for the time being? I think I'd like to proceed with this in the following way:

* Merge the PR at https://github.com/backdrop/backdrop/pull/3688
* Entirely suspend the Zen.CI application access on GitHub. This will disable firing of GitHub webhooks to Zen.CI, but we can restore it if needed.
* Open a follow-up issue to clean up Zen.CI from the core repository.

**Comment by indigoxela** [comment]
> Could you restore the Zen.CI scripts for the time being?

Sure, done.

Two random failures this time, known buddies, CommentInterfaceTest and PathPatternBulkUpdateTestCase.

**Comment by quicksketch** [comment]
Awesome! As we're in the middle of the last effort before feature freeze, I think it might be better to wait until after Sept 1 to merge this so we don't cause unexpected problems. Either that or we run both GitHub and Zen.CI for a short while?

**Comment by klonos** [comment]
> Either that or we run both GitHub and Zen.CI for a short while?

Run both I think. That would allow us to have a fallback, plus spot any differences.

**Comment by indigoxela** [comment]
> As we're in the middle of the last effort before feature freeze, I think it might be better to wait until after Sept 1 to merge this

> Run both I think....

I'm fine with either. Waiting another two or three days doesn't hurt, running both in parallel for a short while shouldn't cause trouble, too.

**Comment by klonos** [comment]
> Waiting another two or three days doesn't hurt...

Right! ...I misread that as Sept 15. A few days wouldn't hurt indeed.

**Comment by quicksketch** [comment]
I have merged https://github.com/backdrop/backdrop/pull/3688 into 1.x. Since we don't usually have PRs against other versions than 1.x, I'm not going to cherry-pick this back into 1.19.x or other older branches.

For now I will disable (but not delete) the Zen.CI app on the backdrop repository. Let's continue to work on the (not-so) random failures that we're still seeing with some frequency. I think it's possible that GitHub Actions slower execution time might be causing the random failures to be increasing, as frequency these issues are driven by race conditions or time-sensitive requests.

Great job folks and especially @indigoxela for putting in the vast amount of work here to get this up and running!

**Comment by quicksketch** [comment]
Follow-up task https://github.com/backdrop/backdrop-issues/issues/5197
